Mathlib/RingTheory/Binomial.lean

```diff
@@ -0,0 +1,124 @@
+
+/-!
+# Binomial semirings and binomial rings
+
+In this file we introduce binomial semirings and binomial rings, together with related functions
+like binomial coefficients.
+
+According to our main reference [Elliott] (which lists many equivalent conditions), a binomial ring
+is a torsion-free commutative ring R such that for any `x ∈ R` and any `k ∈ ℕ`, the product
+`x(x-1)⋯(x-k+1)` is divisible by `k!`.  The torsion-free condition lets us divide by `k!`
+unambiguously, so we get uniquely defined binomial coefficients.
+
+The defining condition doesn't require commutativity, and we get a theory with essentially the same
+power by replacing subtraction with addition.  Thus, we consider a semiring R in which
+multiplication by factorials is injective, and demand that for all `x ∈ R` and any `k ∈ ℕ`,
+`x(x+1)⋯(x+(k-1))` is divisible by `k!`.
+The quotient is called `multichoose r k`, following the convention given for natural numbers.
+
+## References
+* [J. Elliott, *Binomial rings, integer-valued polynomials, and λ-rings*][elliott2006binomial]
+
+-/
+
+namespace Ring
+
+universe u
+
+variable {R: Type u}
+
+section Semiring
+
+variable [Semiring R]
+
+/-- pochEval directly evaluates the n-th Pochhammer polynomial at an element r. -/
+def pochEval (r : R) : ℕ → R
+  | 0 => 1
+  | (k + 1) => (r + k) * (pochEval r k)
+
+theorem pochEval_zero (r : R) :
+  pochEval r 0 = 1 := rfl
+
+theorem pochEval_succ (r : R) (k : ℕ) :
+  pochEval r (k.succ) = (r + k) * (pochEval r k) := by rfl
+
+theorem pochEval_cast (n : ℕ) : ∀ (k : ℕ), pochEval (n : R) k = (pochEval n k)
+| 0 => by rw [pochEval_zero, pochEval_zero, Nat.cast_one]
+| (k + 1) => by
+  rw [pochEval_succ, pochEval_succ, Nat.cast_mul, pochEval_cast n k, Nat.cast_add]
+  norm_cast
+
+theorem Commute.mem_comm_translate (r s : R) (n : ℕ) (h: Commute r s) : Commute r (s + n) := by
+  refine Commute.add_right h (Nat.commute_cast r n)
+
+theorem Commute.translate_comm_mem (r s : R) (n : ℕ) (h : Commute r s) : Commute (r + n) s := by
+  refine Commute.add_left h (Nat.cast_commute n s)
+
+theorem Commute.translates_comm (r s : R) (k n : ℕ) (h : Commute r s): Commute (r + k) (s + n) := by
+  refine Commute.add_right ?_ ?_
+  refine Commute.translate_comm_mem r s k h
+  exact Nat.commute_cast (r+k) n
+
+theorem Commute.translate_comm_pochEval (r s : R) (k : ℕ) (h : Commute r s) : ∀ (n : ℕ),
+  Commute (r + k) (pochEval s n)
+| 0 => by
+  rw [pochEval_zero]
+  exact Commute.one_right (r + ↑k)
+| (n + 1) => by
+  rw [pochEval_succ]
+  refine Commute.mul_right ?_ ?_
+  refine Commute.translates_comm r s k n h
+  exact Commute.translate_comm_pochEval r s k h n
+
+theorem pochEval_eq_pochhammer_eval (r : R) :
+  ∀ (k : ℕ), pochEval r k = Polynomial.eval r (pochhammer R k)
+| 0 => by rw [pochhammer_zero, Polynomial.eval_one, pochEval_zero]
+| (k + 1) => by
+  rw [pochhammer_succ_eval, ← pochEval_eq_pochhammer_eval r k, pochEval_succ,
+    Commute.translate_comm_pochEval]
+  rfl
```

Comment:
I tried defining `pochEval` (or `ascPochEval`) using polynomial evaluation, but it the compiler complains unless I add `noncomputable`.  I thought it would be nice if the binomial coefficient functions (when eventually defined) were computable.  (Is there an advantage to having noncomputable defs?)

Reply:
Okay, I've changed my mind, and have defined ascPochEval as you suggested.  The lets me eliminate the comparison theorem.

Reply:
Sorry, I've changed my mind again to make a recursive definition, and changed the name to `Ring.ascFactorial`.